packages/SwingSet/src/vats/comms/delivery.js

```diff
@@ -125,41 +141,56 @@ export function makeDeliveryKit(state, syscall, transmit, clistKit, stateKit) {
   }
 
   function resolveFromRemote(remoteID, message) {
-    const sci = message.indexOf(';');
-    assert(sci !== -1, details`missing semicolon in resolve ${message}`);
-    // message is created by resolveToRemote, so one of:
-    // `resolve:object:${target}:${resolutionRef};`
-    // `resolve:data:${target}${rmss};${resolution.body}`
-    // `resolve:reject:${target}${rmss};${resolution.body}`
-
-    const pieces = message.slice(0, sci).split(':');
-    // pieces[0] is 'resolve'
-    const type = pieces[1];
-    const remoteTarget = pieces[2];
-    const remoteSlots = pieces.slice(3); // length=1 for resolve:object
-    insistRemoteType('promise', remoteTarget); // slots[0] is 'rp+NN`.
-    const vpid = getLocalForRemote(remoteID, remoteTarget);
-    // rp+NN maps to target=p-+NN and we look at the promiseTable to make
-    // sure it's in the right state.
-    insistPromiseIsUnresolved(vpid);
-    insistDeciderIsRemote(vpid, remoteID);
-
-    const slots = remoteSlots.map(s => provideLocalForRemote(remoteID, s));
-    const body = message.slice(sci + 1);
-    const data = harden({ body, slots });
-    changeDeciderFromRemoteToComms(vpid, remoteID);
+    // message is created by resolveToRemote.  It consists of 1 or more
+    // resolutions, separated by newlines, each taking the form:
+    // `resolve:${rejected}:${target}${rmss};${resolution.body}`
+    const subMessages = message.split('\n');
+    const resolutions = [];
+    for (const submsg of subMessages) {
+      const sci = submsg.indexOf(';');
+      assert(sci !== -1, details`missing semicolon in resolve ${submsg}`);
+
+      const pieces = submsg.slice(0, sci).split(':');
+      assert(pieces[0] === 'resolve');
+      const rejected = pieces[1] === 'reject';
+      assert(rejected || pieces[1] === 'fulfill');
+      const remoteTarget = pieces[2];
+      const remoteSlots = pieces.slice(3); // length=1 for resolve:object
+      insistRemoteType('promise', remoteTarget); // slots[0] is 'rp+NN`.
+      const vpid = getLocalForRemote(remoteID, remoteTarget);
+      // rp+NN maps to target=p-+NN and we look at the promiseTable to make
+      // sure it's in the right state.
+      insistPromiseIsUnresolved(vpid);
+      insistDeciderIsRemote(vpid, remoteID);
+
+      const slots = remoteSlots.map(s => provideLocalForRemote(remoteID, s));
+      const body = submsg.slice(sci + 1);
+      const data = harden({ body, slots });
+      changeDeciderFromRemoteToComms(vpid, remoteID);
+      resolutions.push([vpid, { rejected, data }]);
```

Comment:
It might be nice to have `resolutions` take the same shape as we do everywhere else: a three-element array. But I'm happy to see this land as-is.

Oh, wait, I see that at least the kernel-to-liveslots pathway also uses this shape. So maybe we're converging on `[ vpid, { rejected, data } ]` instead of `[ vpid, rejected, data ]`, which is fine. (I'm marginally in favor of the three-element array, but it's because I'm concerned about the hard-to-diagnose `undefined` we get when we misspell something, and we could use type annotations to help discover things like that).

Reply:
This was one of the "you push in down here, it pops up over there, you push it down there and it pops up over _there_" things that was hard to get all coordinated, so I'm reluctant to tackle this right now.  That said, having a uniform representation is the right thing to do, so I've filed this as issue #2248